clang-tools-extra/clangd/unittests/InlayHintTests.cpp

```diff
@@ -1637,6 +1678,144 @@ TEST(TypeHints, SubstTemplateParameterAliases) {
                         ExpectedHint{": static_vector<int>", "vector_name"});
 }
 
+template <typename... Labels>
+void assertTypeLinkHints(StringRef Code, StringRef HintRange,
+                         Labels... ExpectedLabels) {
+  Annotations Source(Code);
+  auto HintAt = [&](llvm::ArrayRef<InlayHint> InlayHints,
+                    llvm::StringRef Range) {
+    auto *Hint = llvm::find_if(InlayHints, [&](const InlayHint &InlayHint) {
+      return InlayHint.range == Source.range(Range);
+    });
+    assert(Hint && "No range was found");
+    return llvm::ArrayRef(Hint->label);
+  };
+
+  TestTU TU = TestTU::withCode(Source.code());
+  TU.ExtraArgs.push_back("-std=c++2c");
+  auto AST = TU.build();
+
+  Config C;
+  C.InlayHints.TypeNameLimit = 0;
+  WithContextValue WithCfg(Config::Key, std::move(C));
+
+  auto Hints = hintsOfKind(AST, InlayHintKind::Type);
+  EXPECT_THAT(HintAt(Hints, HintRange),
+              ElementsAre(HintLabelPieceMatcher(ExpectedLabels, Source)...));
+}
+
+TEST(TypeHints, Links) {
+  StringRef Source(R"cpp(
+    $Package[[template <class T, class U>
```

Comment:
Is there a reason to prefer putting full decl source ranges into `InlayHintLabelPart.location`, rather than single-token source ranges (e.g. just the range `Package`) like in the go-to-definition response?

What vscode seems to do is call `textDocument/definition` at the start position of the range, and navigate to the resulting definition location.

I guess most of the time that works either with a full source range or a single-token source range as input, but I think a single-token source range might be more robust (less chance for edge cases where the first token of the full source range is claimed by some descendant node, and also less chance for edge cases where the full source range spans macro expansion boundaries).

We have [some utilities](https://searchfox.org/llvm/rev/1664610130b88ef168e33eddfe973a3f11bd4261/clang-tools-extra/clangd/XRefs.cpp#414-415) for getting the single-token source range for the go-to-def response, which we could probably reuse.

(As a bonus, it makes the tests a bit easier to write and read.)

Reply:
> Is there a reason to prefer putting full decl source ranges into InlayHintLabelPart.location, rather than single-token source ranges (e.g. just the range Package) like in the go-to-definition response?

Frankly, it is because the protocol requires a source range of something, and I'm concerned that providing only one source location doesn't really make a better experience. But I have to say I've never attempted it before, so I think it's worth a try.

> What vscode seems to do is call textDocument/definition at the start position of the range, and navigate to the resulting definition location.

I have realized that problem; it looks like a call to `textDocument/definition` is not cheap and it usually takes a few seconds to respond (taking us to the definition/getting the hover content) on my other large project.

> We have [some utilities](https://searchfox.org/llvm/rev/1664610130b88ef168e33eddfe973a3f11bd4261/clang-tools-extra/clangd/XRefs.cpp#414-415) for getting the single-token source range for the go-to-def response, which we could probably reuse.
> (As a bonus, it makes the tests a bit easier to write and read.)

Thanks! I really appreciate that and will explore it soon. :)

Reply:
@HighCommander4 I explored it a bit more:

I explored it a bit more:

1. Using `RecordDecl::getBeginLoc()` does work in terms of the protocol, but this SourceLocation points to the first keyword token of the definition, i.e., the location of struct/class, which doesn't seem acceptable to us.

2. The better approach to having the identifier location is to resort to `TypeLocs`, which is involved regarding the current implementation of `addTypeHint()` because it's taking a `QualType` as a parameter and changing it to `TypeSourceInfo/TypeLoc` might require some surgery.

I think I'm in favor of shifting the current `TypeHint` implementation to rely on `TypeLocs`. We probably need a separate patch before this one. WDYT? Or is there any possible approach to obtaining the location of the identifier without using `TypeLocs`?

Reply:
>    1. Using `RecordDecl::getBeginLoc()` does work in terms of the protocol, but this SourceLocation points to the first keyword token of the definition, i.e., the location of `struct/class`, which doesn't seem acceptable to us.

I was thinking of `Decl::getLocation()`, which should point to the location of the declaration's name.

Reply:
Oh, thanks - indeed we have an identifier location with `getLocation()`.
(I promise I have seen `getLocation()` while looking at the definition of `Decl::getBeginLoc()`, but I thought it was the same as `getLocation()` because that's how the base `getSourceRange()` implements, although it is overridden by the subclasses. I should have looked into the subclasses' constructors, my fault!)